libs/@local/hashql/ast/src/lowering/special_form_expander/error.rs

```diff
@@ -1110,3 +1131,75 @@ pub(crate) fn duplicate_closure_generic(
 
     diagnostic
 }
+
+pub(crate) fn field_literal_type_annotation(span: SpanId) -> SpecialFormExpanderDiagnostic {
+    let mut diagnostic = Diagnostic::new(
+        SpecialFormExpanderDiagnosticCategory::FieldLiteralTypeAnnotation,
+        Severity::Error,
+    );
+
+    diagnostic
+        .labels
+        .push(Label::new(span, "Remove this type annotation"));
+
+    diagnostic.add_help(Help::new(
+        "Field access using numeric literals cannot have type annotations. The literal represents \
+         the field index directly and doesn't need a separate type.",
+    ));
+
+    diagnostic.add_note(Note::new(
+        "When using numeric literals for field access (e.g., in tuple destructuring), the number \
+         itself indicates the position and cannot be annotated with a type.",
+    ));
+
+    diagnostic
+}
+
+pub(crate) fn invalid_field_literal_type(span: SpanId) -> SpecialFormExpanderDiagnostic {
+    let mut diagnostic = Diagnostic::new(
+        SpecialFormExpanderDiagnosticCategory::InvalidFieldLiteralType,
+        Severity::Error,
+    );
+
+    diagnostic
+        .labels
+        .push(Label::new(span, "Use an integer literal here"));
+
+    diagnostic.add_help(Help::new(
+        "Field access using literals requires an integer literal that specifies the field index. \
+         Other literal types like strings, booleans, or numbers cannot be used for field indexing.",
+    ));
+
+    diagnostic.add_note(Note::new(
+        "Valid field access examples:\n- (access tuple 0) - accesses first field\n- (access \
+         record 2) - accesses third field\n- (access data my_field) - accesses named field",
+    ));
+
+    diagnostic
+}
+
+pub(crate) fn field_index_out_of_bounds(span: SpanId) -> SpecialFormExpanderDiagnostic {
+    let mut diagnostic = Diagnostic::new(
+        SpecialFormExpanderDiagnosticCategory::FieldIndexOutOfBounds,
+        Severity::Error,
+    );
+
+    diagnostic.labels.push(Label::new(
+        span,
+        "Use a valid field index within usize bounds",
+    ));
+
+    diagnostic.add_help(Help::new(
+        "Field indices must be non-negative integers that fit within the bounds of usize. Very \
```

Comment:
`size` might be correct, but it differs between platforms. Do we actually parse it as `usize` or as `u64`?

Reply:
We always parse it as usize, because we still use ordinary collections, if we were to parse it as u64, on 32-bit platforms we'd crash when we try to index (or just couldn't index)

Reply:
I guess that's fine. If someone needs bigger index than 4B on a 32 bit system they have different issues 😅